// @juspay-tech/react-hyper-js/src/Context.bs.js
// Generated by ReScript, PLEASE EDIT WITH CARE

import * as React from "react";
import * as OrcaJs from "./OrcaJs.bs.js";
import * as Js_dict from "rescript/lib/es6/js_dict.js";
import * as Js_json from "rescript/lib/es6/js_json.js";
import * as Belt_Option from "rescript/lib/es6/belt_Option.js";

function confirmPaymentFn(_elements) {
  return Promise.resolve({});
}

function confirmCardPaymentFn(_clientSecretId, _data, _options) {
  return Promise.resolve({});
}

function retrievePaymentIntentFn(_paymentIntentId) {
  return Promise.resolve({});
}

function paymentRequest(options) {
  return options;
}

var defaultSwitchContext = {
  clientSecret: "",
  confirmPayment: confirmPaymentFn,
  confirmCardPayment: confirmCardPaymentFn,
  retrievePaymentIntent: retrievePaymentIntentFn,
  paymentRequest: paymentRequest
};

var switchContext = React.createContext(defaultSwitchContext);

function makeProps(value, children, param) {
  return {
          value: value,
          children: children
        };
}

var make = switchContext.Provider;

var SwitchContextProvider = {
  makeProps: makeProps,
  make: make
};

function getString(dict, key, defaultVal) {
  return Belt_Option.getWithDefault(Belt_Option.flatMap(Js_dict.get(dict, key), Js_json.decodeString), defaultVal);
}

function elementsOptionObjMapper(options) {
  var dict = Belt_Option.getWithDefault(Js_json.decodeObject(options), {});
  return {
          fonts: Belt_Option.getWithDefault(Belt_Option.flatMap(Js_dict.get(dict, "fonts"), Js_json.decodeArray), []),
          locale: getString(dict, "locale", ""),
          clientSecret: getString(dict, "clientSecret", ""),
          appearance: Belt_Option.getWithDefault(Belt_Option.flatMap(Js_dict.get(dict, "appearance"), Js_json.decodeObject), {}),
          loader: getString(dict, "loader", "auto")
        };
}

function update(_options) {
  
}

function getElement(_componentName) {
  
}

function fetchUpdates(param) {
  return new Promise((function (resolve, param) {
                setTimeout((function (param) {
                        resolve({});
                      }), 1000);
              }));
}

function create(_componentType, _options) {
  return OrcaJs.defaultPaymentElement;
}

var defaultElementsContext_options = {
  fonts: [],
  locale: "",
  clientSecret: "",
  appearance: {},
  loader: ""
};

var defaultElementsContext = {
  options: defaultElementsContext_options,
  update: update,
  getElement: getElement,
  fetchUpdates: fetchUpdates,
  create: create
};

var elementsContext = React.createContext(defaultElementsContext);

function makeProps$1(value, children, param) {
  return {
          value: value,
          children: children
        };
}

var make$1 = elementsContext.Provider;

var ElementsContextProvider = {
  makeProps: makeProps$1,
  make: make$1
};

export {
  confirmPaymentFn ,
  confirmCardPaymentFn ,
  retrievePaymentIntentFn ,
  paymentRequest ,
  defaultSwitchContext ,
  switchContext ,
  SwitchContextProvider ,
  getString ,
  elementsOptionObjMapper ,
  update ,
  getElement ,
  fetchUpdates ,
  create ,
  defaultElementsContext ,
  elementsContext ,
  ElementsContextProvider ,
}
/* confirmCardPaymentFn Not a pure module */
